src/localisation/include/EKF.h

```diff
@@ -0,0 +1,121 @@
+    // matrices
+    Matrix3d p; // covariance matrix
+    Matrix3d j; // jacobian matrix (differentiate angle)
+    Matrix3d q; // used to calculate the covariance matrix (process model)
+    Matrix3d s; // used to calculate the Kalman gain
+    Matrix3d k; // Kalman gain
+    Vector3d position; // vector of bot position
+    Vector3d z; // vector of gps and imu position/orientation measurements
+    MatrixXd ju{MatrixXd(3, 2)}; // jacobian matrix (differentiate speed)
+    // const matrices
+    // accounts for unknown external influences
+    Matrix3d q1{(Matrix3d() << 0.01 * 0.01,
+                 0,
+                 0,
+                 0,
+                 0.01 * 0.01,
+                 0,
+                 0,
+                 0,
+                 (1. * (M_PI / 180.)) * (1. * (M_PI / 180.)))
+                .finished()};
+    // covariance of the encoder speed measurement and the imu angular velocity
+    // measurement
+    Matrix2d pu{
+    (Matrix2d() << sdev_speed * sdev_speed, 0, 0, sdev_gyro* sdev_gyro)
+    .finished()};
+    // observation matrix
+    Matrix3d h{(Matrix3d() << 1, 0, 0, 0, 1, 0, 0, 0, 1).finished()};
+    // covariance of the gps position measurements and the imu orientation
+    // measurement
+    Matrix3d r{(Matrix3d() << sdev_gps_x_measurement * sdev_gps_x_measurement,
+                0,
+                0,
+                0,
+                sdev_gps_y_measurement* sdev_gps_y_measurement,
+                0,
+                0,
+                0,
+                sdev_angle_measurement* sdev_angle_measurement)
+               .finished()};
+    // covariance for initial position currently assume it's perfect
+    Matrix3d intial_p{(Matrix3d() << 0, 0, 0, 0, 0, 0, 0, 0, 0).finished()};
+
+  public:
+    // ros's specific position data type
+    geometry_msgs::Pose bot_position;
+    // Default Constructor
+    EKF();
+
+    // Constructor
+    EKF(double pos_x, double pos_y, double pos_z, double yaw_angle);
+
+    /**
+     * Takes an angle and adjusts it if necessary so that it says within the
+     * bounds of -pi and +pi
+     *
+     * @param an angle in radians
+     *
+     * @return a rebounded angle in radians
+     */
+    static double defineAngleInBounds(double angle);
+
+    /**
+     * Takes in encoder and imu measurements to perform the process model of the
+     * kalman filter.
+     *
+     * @param ros specific data types for encoder and imu measurement data
+     */
+    void
+    processModel(nav_msgs::Odometry encoder, sensor_msgs::Imu imu, double dt);
+
+    /**
+     * Takes in gps and imu measurements to perform the measurement model of the
+     * kalman filter.
+     *
+     * @param ros specific data types for gps and imu measurement data
+     *
+     * @return ros specific data type of the bot's position
+     */
+    geometry_msgs::Pose measurementModel(nav_msgs::Odometry gps,
+                                         sensor_msgs::Imu imu);
```

Comment:
Since people not familiar with EKF's will probably be using this, these might be better named `addMeasurement` and `estimateState` (rather then `processModel` and `measurementModel` respectively).

Also, would it be possible to have to have two `addMeasurement` functions (one for encoder readings, and one for GPS readings), and a `getState` function? I just think that it might be a bit more intuitive for those not familiar with EKF's. (Though you've more experience with EKF's then I, so feel free to let me know if I'm completely off the mark here with how it works)

Reply:
best to split it into two steps the estimate state step and the update state step.